caom2pipe/manage_composable.py

```diff
@@ -2924,3 +2925,146 @@ def find_missing(compare_this, to_this):
+                         # than one replacement value
+
+    Caveats:
+
+    Some CAOM2 attributes are not assignable. These attributes cannot be
+    repaired using this class.
+
+    List content cannot be repaired using this class.
+
+    Use the string value 'none' to set an attribute to None.
+
+    Use '*' to replace part of a string only works with str values, and only
+    trailing changes, because there haven't been any preceding cases that I've
+    needed to write code for yet.
+    """
+
+    VALUE_REPAIR = 'value_repair'
+
+    def __init__(self):
+        super(ValueRepairCache, self).__init__()
+        self._value_repair = self.get_from(ValueRepairCache.VALUE_REPAIR)
+        self._key = None
+        self._values = None
+        self._logger = logging.getLogger(self.__class__.__name__)
+
+    def repair(self, observation):
+        try:
+            for key, values in self._value_repair.items():
+                self._logger.debug(f'Checking for {key}')
+                self._key = key
+                self._values = values
+                attribute_name = None
+                entity_name = None
+                bits = key.split('.')
+                if len(bits) == 2:
+                    entity_name = bits[0]
+                    attribute_name = bits[1]
+                elif len(bits) == 3:
+                    entity_name = bits[1]
+                    attribute_name = bits[2]
+                # else:
+                # the other possible lengths are for chunk entity/attribute
+                # checking
+
+                if key.startswith('observation'):
+                    self._find_and_fix(observation, entity_name,
+                                       attribute_name)
+                elif key.startswith('plane'):
+                    for plane in observation.planes.values():
+                        self._find_and_fix(plane, entity_name, attribute_name)
+                elif key.startswith('artifact'):
+                    for plane in observation.planes.values():
+                        for artifact in plane.artifacts.values():
+                            self._find_and_fix(artifact, entity_name,
+                                               attribute_name)
+                elif key.startswith('part'):
+                    for plane in observation.planes.values():
+                        for artifact in plane.artifacts.values():
+                            for part in artifact.parts.values():
+                                self._find_and_fix(part, entity_name,
+                                                   attribute_name)
+                elif key.startswith('chunk'):
+                    for plane in observation.planes.values():
+                        for artifact in plane.artifacts.values():
+                            for part in artifact.parts.values():
+                                for chunk in part.chunks:
+                                    self._recurse(chunk, 'chunk', bits[1:])
```

Comment:
Not sure why `chunk` is treaded differently. To me, this is the most generic case and can be applied to all the other components.

Reply:
Done.